File: Python/leetcode 15.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for i in range(len(nums)-2):
            if i == 0 or (i > 0 and nums[i] != nums[i-1]):
                self.twoSum(res, i, nums)
        return res    
    
    def twoSum(self, res, idx, nums):
        target = -nums[idx]
        lo, hi = idx + 1, len(nums) - 1
        while (lo < hi):
            if nums[lo] + nums[hi] == target:
                res.append((nums[idx], nums[lo], nums[hi]))
                while (lo < hi) and (nums[lo] == nums[lo+1]):
                    lo += 1
                while (lo < hi) and (nums[hi] == nums[hi-1]):
                    hi -= 1
                lo += 1
                hi -= 1
            elif nums[lo] + nums[hi] < target:
                lo += 1
            else:
                hi -= 1
    


    
        
        

File: Java/leetcode 15.java
//Solution1: Two pointers T: O(n^2) S: O(1)
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums == null || nums.length == 0) return res;
        Arrays.sort(nums);
        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i-1]) continue;
            List<List<Integer>> tmp = twoSum(nums, i);
            if (tmp.size() > 0) {
                res.addAll(tmp);
            }
        }
        return res;
    }

    public List<List<Integer>> twoSum(int[] nums, int idx) {
        int target = -nums[idx];
        int lo = idx + 1, hi = nums.length - 1;
        List<List<Integer>> res = new ArrayList<>(3);
        while (lo < hi) {
            int sum = nums[lo] + nums[hi];
            if (sum < target) lo++;
            else if (sum > target) hi--;
            else {
                res.add(Arrays.asList(nums[idx], nums[lo], nums[hi]));
                while (lo < hi && nums[lo] == nums[lo+1]) lo++;
                while (lo < hi && nums[hi] == nums[hi-1]) hi--;
                lo++;
                hi--;
            }
        }
        return res;
    }
}